### DISCUSSION TITLE
Re-initialize component after page transition

### DISCUSSION THREAD

**Discussion by WingsOfFury** [swup/swup#817] [Help] [answered]
Hello everyone. After I updated all the dependencies of Astro and Swup, the toggle switch started having problems. After every refresh I can use the page but as soon as I change pages it stops working. Maybe someone here can help me?  If I delete the ID="swup" in main then the toggle switch works again but the fade transition doesn't work smoothly. Rather choppy.Regards.

https://github.com/swup/swup/assets/85767977/b74fdbba-5dd5-41dc-8772-9ea5bf49633c

<img width="824" alt="Bildschirmfoto 2023-11-11 um 22 33 23" src="https://github.com/swup/swup/assets/85767977/54c51585-6b66-4941-a713-e1a46067dffa">
<img width="775" alt="Bildschirmfoto 2023-11-11 um 22 33 05" src="https://github.com/swup/swup/assets/85767977/2d9992ee-5c4b-4fc6-9d25-6151b18a2882">
<img width="584" alt="Bildschirmfoto 2023-11-11 um 22 32 56" src="https://github.com/swup/swup/assets/85767977/2d4dca21-df40-43eb-8c76-832f68b97e70">
<img width="601" alt="Bildschirmfoto 2023-11-11 um 22 34 40" src="https://github.com/swup/swup/assets/85767977/607be75c-da0f-4f21-ac71-a13313ee7e84">

**Comment by daun** [comment]
Which version did you upgrade from, and which version did you upgrade to? If upgrading from swup 3 to 4, the [official upgrade guide](https://swup.js.org/getting-started/upgrading/) might give a clue here, in case it has to do with updated hook names.

In any case, it'd be helpful if you could post your whole JS code here, including the part that manages the dark mode toggle.

Any console errors?

**Comment by WingsOfFury** [comment]
Hello Phillip. Thanks for the quick reply. I updated from 3 to 4. After the plugins had been on strike for a bit, I completely deleted Swup with the plugins (npm uninstall) and installed the plugins with (npm install swup @swup/scroll-plugin and so on). Thanks for the tip about the hooks. I was so busy with Astro that I missed the hooks. I added the hooks but unfortunately something is wrong. When it comes to plugins, “SwupScrollPlugin” is strangely colored blue instead of yellow. I placed the swup.hook.before directly above the //Swup Page Transition but unfortunately this didn't work either. Sure, no problem with the pictures. Regards.
<img width="402" alt="Bildschirmfoto 2023-11-12 um 17 03 28" src="https://github.com/swup/swup/assets/85767977/1c1a6115-ddd7-4cf6-a259-626912e9c446">
<img width="1402" alt="Bildschirmfoto 2023-11-12 um 17 03 55" src="https://github.com/swup/swup/assets/85767977/7853aaee-4c73-4cdf-af4f-4f3bed9e2e1c">
<img width="1387" alt="Bildschirmfoto 2023-11-12 um 17 04 14" src="https://github.com/swup/swup/assets/85767977/500df6f0-9050-4f23-ae9f-080155c12369">

**Reply by daun** [reply]
To re-run your init code on page change, the hook on line 122 shouldn't be empty but hold a reference to `init`:

```js
swup.hooks.on('content:replace', () => init());
```

**Reply by WingsOfFury** [reply]
Unfortunately without an effect.

<img width="837" alt="Bildschirmfoto 2023-11-13 um 00 01 57" src="https://github.com/swup/swup/assets/85767977/bde58f1e-bb91-4ba7-81e7-a31779c34eba">

**Comment by WingsOfFury** [comment]
This is currently in the console.
<img width="1037" alt="Bildschirmfoto 2023-11-12 um 21 27 15" src="https://github.com/swup/swup/assets/85767977/05185ee8-9cec-450a-9249-05e044f63848">

**Reply by daun** [reply]
You're creating swup on line 6 and then again on line 112. This will be the issue here.

**Reply by WingsOfFury** [reply]
Yes, that's totally strange. If I delete var swup the double declaration in the console is gone but unfortunately toggleSwitch still doesn't work.

<img width="878" alt="Bildschirmfoto 2023-11-12 um 23 47 36" src="https://github.com/swup/swup/assets/85767977/f583dfb0-358f-4650-b307-098d7d410849">
<img width="979" alt="Bildschirmfoto 2023-11-12 um 23 50 46" src="https://github.com/swup/swup/assets/85767977/884ba7da-5a88-43d6-abfa-fa73e40589f9">

**Reply by daun** [reply]
It will still lead to problems as you're creating two Swup instances, which is unsupported/undefined behavior. You need to make sure you're only ever running `new Swup()` once in your whole JS, i.e. remove the one on line 6.

**Reply by WingsOfFury** [reply]
Understood. is deleted. Now this error message appears.

<img width="1152" alt="Bildschirmfoto 2023-11-13 um 00 13 46" src="https://github.com/swup/swup/assets/85767977/c2c6ab8e-f458-49a0-ae48-d0a1baff760c">
<img width="1067" alt="Bildschirmfoto 2023-11-13 um 00 15 33" src="https://github.com/swup/swup/assets/85767977/4293f253-ec53-4965-a4ff-ade2c0951a34">

**Reply by daun** [reply]
Sure, that's to be expected. You're accessing `swup.hooks.before`, and there is no variable named `swup`. I don't know why you need to create an empty hook handler on that line, removing it should solve this. If you need the `swup` instance variable earlier, you probably need to move the setup code higher up.

**Comment by WingsOfFury** [comment]
Here is a picture of the Astro.config.mjs.

<img width="732" alt="Bildschirmfoto 2023-11-13 um 00 09 06" src="https://github.com/swup/swup/assets/85767977/f93cb0f8-326e-47e3-838b-41cd0f64bc4a">

**Reply by daun** [reply]
You'll only ever need to do one or the other: manual setup or astro integration. Using both will create two separate swup instances, leading to problems.

**Reply by WingsOfFury** [reply, marked as answer]
It's worked out. I thought I also have to swup.hooks.before("content:replace", () => {} ); maintain it so that it takes effect. Oh man. Difficult birth. With Init at swup.hooks.on is also correct because if I leave out init the dark mode doesn't work. I thank you from the bottom of my heart for your help and your patience, my good friend. Now I can go to sleep with peace of mind. I wish you a good night too.

https://github.com/swup/swup/assets/85767977/bfa0f2e2-77bf-446a-8beb-dca4ef3ffa39

**Reply by daun** [reply]
Glad you could solve it!
